Fix panic during the expansion of `column!`

marcogroppo opened the pull request:
Fixes #2379. Well, this isn't the "proper" fix but it doesn't hurt, IMHO.

The problem is that `to_col_number`, called by `column_expand`, receives a position number that isn't included in the text range of the file. My (very limited) understanding is that the text is the one of the original file, while `pos` is relative to something else, probably the text of the macro. Notice that in this case the `column!` expansion seems to be triggered by `assert_eq!`, so we're in the middle of another expansion. This PR simply avoids the panic by checking the length of the text.

r? @edwin0cheng

edwin0cheng commented:
Yes, I think your explanation is correct. Maybe we could use the new `Expansion` api the remap the pos to original file but it is a little tricky since `to_col_number` is live behind the `salsa DB` firewall. But I think at this moment your bound check is okay..

But would you mind to fix `to_line_number` too? It has the same problem.

edwin0cheng commented on crates/ra_hir_expand/src/builtin_macro.rs:
```diff
@@ -114,20 +114,25 @@ fn stringify_expand(
-    let mut col_num = 1;
 
-    for c in text[..pos.to_usize()].chars().rev() {
+    let pos = pos.to_usize();
```
Would you mind to add a `FIXME` which indicate this code is just a hot fix?

marcogroppo commented:
> But would you mind to fix `to_line_number` too? It has the same problem.

I didn't fix it because it uses an iterator, so it shouldn't panic. It could have returned the wrong line number, tough... I've added the same check

flodiebold commented:
I wonder, why are we even trying to expand this to the correct value? It doesn't make a difference, does it? Couldn't we just always put `0` there?

marcogroppo commented:
@flodiebold mmm ok, so you think it would be better to just return 0 both for `column!` and `line!`?

lnicola commented:
It's visible when using the new "expand macro" command, I guess.

marcogroppo commented:
> It's visible when using the new "expand macro" command, I guess.

No, it isn't. I don't know why, but if you try to expand `assert_eq!` using the RA command you see `$crate::column!()` and `$crate::line!()`.
I've modified the PR so that both `line!` and `column!` just return `0` in case of errors and I've added a `FIXME`.

kjeremy commented:
> 
> 
> It's still visible when using the new "expand macro" command, right?

If expanding `column!` it is. for `assert_eq!` there's some more work that needs to be done unrelated to `line!` and `column!`

edwin0cheng commented:
> and for line!, isn't it even harmful towards incrementality?)

@flodiebold If I understand correctly,  it should be no difference. Even in normal macro call, If the line number changed, `MacroCallId` will be changed too (i.e `MacroCallId contains `ast::MacroCall` which related to file position).

edwin0cheng commented:
> using the RA command you see $crate::column!() and $crate::line!()

Maybe we didn't handle `Hygiene` properly in `SourceAnalyser` yet ? I will investigate that later. :)

and LGTM now, thanks.
bors r+

flodiebold commented:
> @flodiebold If I understand correctly, it should be no difference. Even in normal macro call, If the line number changed, `MacroCallId` will be changed too (i.e `MacroCallId contains `ast::MacroCall` which related to file position).

It'd still make a difference for later steps like type inference though :slightly_smiling_face:

edwin0cheng commented:
Oh... you are right But I can imagine some "good" c++ programmer to use the following trick to generate a compile time random number:

```rust
// pre-generate a list of  data/hash[*].txt  of random number
macro_rules compile_time_random {
  () => ( include_str!(concat!("data/hash", line!(), ".txt"))  }
}
```
